Samples/AudioCreation/cs/AudioCreation/Scenario7_RecordPlayback.xaml.cs

```diff
@@ -0,0 +1,570 @@
+
+                HandleOutgoingAudio(requiredSamples);
+
+                /*
+                int currentStatusMsec = DateTime.Now.Millisecond;
+                // update status every tenth of a second
+                if (lastStatusMsec % 100 > currentStatusMsec % 100)
+                {
+                    CoreApplication.MainView.CoreWindow.Dispatcher.TryRunAsync(
+                        Windows.UI.Core.CoreDispatcherPriority.Normal,
+                        UpdateStatusWhilePlaying);
+                }
+                lastStatusMsec = currentStatusMsec;
+                */
+            }
+        }
+
+        // We keep a 1MB AudioFrame and reuse it.
+        static AudioFrame s_audioFrame = new Windows.Media.AudioFrame(1024 * 1024);
+        private int s_zeroByteOutgoingFrameCount;
+
+        /// <summary>
+        /// Handle a frame of outgoing audio to the output device.
+        /// </summary>
+        /// <remarks>
+        /// This pushes a megabyte of audio each time, to greatly reduce the number of callbacks.
+        /// 
+        /// We don't need the requiredSamples argument but we don't 
+        /// </remarks>
+        private unsafe void HandleOutgoingAudio(uint requiredSamples)
+        {
+            if (requiredSamples == 0)
+            {
+                s_zeroByteOutgoingFrameCount++;
+                return;
+            }
+
+            using (AudioBuffer buffer = s_audioFrame.LockBuffer(AudioBufferAccessMode.Write))
+            using (IMemoryBufferReference reference = buffer.CreateReference())
+            {
+                byte* dataInBytes;
+                uint capacityInBytes;
+
+                // Get the buffer from the AudioFrame
+                ((IMemoryBufferByteAccess)reference).GetBuffer(out dataInBytes, out capacityInBytes);
+
+                uint bytesRemaining = capacityInBytes;
+                uint position = 0;
+
+                while (bytesRemaining > 0)
+                {
+                    uint bytesInThisIteration = bytesRemaining;
+                    if (currentIndex + bytesRemaining > maxIndex)
+                    {
+                        bytesInThisIteration = maxIndex - currentIndex;
+                        if (bytesInThisIteration == 0)
+                        {
+                            // wraparound; next iteration will get more bytes
+                            currentIndex = 0;
+                            continue;
+                        }
+                    }
+
+                    fixed (byte* buf = byteBuffer)
+                    {
+                        for (int i = 0; i < bytesInThisIteration; i++)
+                        {
+                            dataInBytes[position + i] = buf[i + currentIndex];
+                        }
+                    }
+                    currentIndex += bytesInThisIteration;
+                    position += bytesInThisIteration;
+
+                    bytesRemaining -= bytesInThisIteration;
```

Comment:
Since you know the data is in multiples of 8-bytes, you could use int64 or at least int32 when looping through the memory buffers, for better efficiency. In the product code we would have used memcpy or CopyMemory (I don't normally program in C# -- are these functions available?) for better efficiency. Lacking memcpy/CopyMemory, we would have looped like you do, but using int64 or inte32 variables. I know the type casting will cause the sample code to look uglier, but it is also important for the sample code to show how to write efficient code. This optimization can be applied to the HandleInput function, above, as well.